### PR TITLE
 Data race on _pplx_g_sched_t::get_scheduler() (#1085)

### PR DESCRIPTION
**Pull request by dimarusyy**
* Make singleton thread-safe

### CODE REVIEW

File: Release/src/pplx/pplx.cpp [hunk 1]
```diff
@@ -63,58 +64,34 @@ static struct _pplx_g_sched_t
...
+            ::pplx::details::_Scoped_spin_lock lock(m_spinlock);
+            sptr = std::atomic_load_explicit(&m_scheduler, std::memory_order_relaxed);
+            if (!sptr)
+            {
+                sptr = std::make_shared<::pplx::default_scheduler_t>();
+                std::atomic_store_explicit(&m_scheduler, sptr, std::memory_order_release);
+            }
         }
+
+        return m_scheduler;
```
Comment (BillyONeal) [inline Release/src/pplx/pplx.cpp]: Bug, needs to return sptr.

File: Release/src/pplx/pplx.cpp [hunk 2]
```diff
@@ -63,58 +64,34 @@ static struct _pplx_g_sched_t
...
 
-    _pplx_g_sched_t() { m_state = post_ctor; }
-
-    ~_pplx_g_sched_t() { m_state = post_dtor; }
-
     sched_ptr get_scheduler()
-    {
-        switch (m_state)
+    {        
+        sched_ptr sptr = std::atomic_load_explicit(&m_scheduler, std::memory_order_consume);
```
Comment (BillyONeal) [inline Release/src/pplx/pplx.cpp]: Not sure how I feel about memory_order_consume; even SG1 thinks that thing is effectively broken. Any objections to changing this to acquire?

File: Release/src/pplx/pplx.cpp [hunk 3]
```diff
@@ -63,58 +64,34 @@ static struct _pplx_g_sched_t
...
-                        m_scheduler = std::make_shared<::pplx::default_scheduler_t>();
-                    }
-                }
-
-                return m_scheduler;
-            default:
-                // This case means the global m_scheduler is not available.
-                // We spin off an individual scheduler instead.
-                return std::make_shared<::pplx::default_scheduler_t>();
+            ::pplx::details::_Scoped_spin_lock lock(m_spinlock);
```
Comment (BillyONeal) [inline Release/src/pplx/pplx.cpp]: As far as I am aware, all the major shared_ptr implementations actually use spinlocks to implement this load operation anyway; perhaps we should just drop the check and take the spinlock every time. It does limit scalability of the system but this shared_ptr city wasn't going to win awards for that anyway.

File: Release/src/pplx/pplx.cpp (hunk 2 above)
Comment (dimarusyy) [reply Release/src/pplx/pplx.cpp]: `consume` and `acquire` should be equivalent here as atomic is not used for synchronization here. Anyway, I'll update with `acquire`.

File: Release/src/pplx/pplx.cpp (hunk 3 above)
Comment (dimarusyy) [reply Release/src/pplx/pplx.cpp]: Am I right that you suggest to drop double-lock here and just acquire spinlock on every call to `get_scheduler()`?

**Comment by BillyONeal** [comment]
Can you look at this alternative and let me know if that makes you happy?

**Comment by BillyONeal** [comment]
I mean this alternative: https://github.com/Microsoft/cpprestsdk/pull/1088

File: Release/src/pplx/pplx.cpp (hunk 3 above)
Comment (BillyONeal) [reply Release/src/pplx/pplx.cpp]: @dimarusyy Yes, that's what my alternative does. Does that solution look acceptable to you?

Comment (dimarusyy) [reply Release/src/pplx/pplx.cpp]: I'm fine as you accepted spinlock acquiring on every call to `get_scheduler()`.  Thanks for comment and update.

**Pull request closed without merging**